recovery/elliptics_recovery/types/dc.py

```diff
@@ -234,18 +264,107 @@ def contains(addresses_with_backends, address, backend_id):
+    appended to the file containing committed keys.
+    If uncommitted key's replica hasn't exceeded prepare timeout, then skip recovering of the key,
+    because the key is under writing and can be committed in the nearest future.
+    '''
+    if ctx.dry_run or ctx.safe:
+        return
+
+    elog = elliptics.Logger(ctx.log_file, int(ctx.log_level))
+    node = elliptics_create_node(address=ctx.address,
+                                 elog=elog,
+                                 wait_timeout=ctx.wait_timeout,
+                                 flags=elliptics.config_flags.no_route_list,
+                                 net_thread_num=1,
+                                 io_thread_num=1,
+                                 remotes=ctx.remotes)
+    session = elliptics.newapi.Session(node)
+    session.trace_id = ctx.trace_id
+    session.exceptions_policy = elliptics.exceptions_policy.no_exceptions
+    session.set_filter(elliptics.filters.all_final)
+    session.ioflags |= elliptics.io_flags.cas_timestamp
+    session.timestamp = ctx.prepare_timeout
+
+    stats = ctx.stats['recover']
+    stats_cmd = ctx.stats['commands']
+
+    for r in results:
+        with open(r.filename, 'ab') as f:
+            for _, batch in groupby(enumerate(load_key_data(r.uncommitted_filename)),
+                                    key=lambda x: x[0] / ctx.batch_size):
+                batch = [item[1] for item in batch]
+                tasks = []
+                for key, key_infos in batch:
+                    for info in key_infos:
+                        if info.flags & elliptics.record_flags.uncommitted:
+                            tasks.append((key, info.group_id, info.size))
+
+                statuses = {} # (key, group_id) -> status
+                for attempt in range(ctx.attempts):
+                    if not tasks:
+                        break
+
+                    batch_sizes = defaultdict(int) # group_id -> batch_size
+                    for _, group_id, key_size in tasks:
+                        batch_sizes[group_id] += key_size
+
+                    timeouts = {group_id: max(60, batch_size / ctx.data_flow_rate)
+                                for group_id, batch_size in batch_sizes.iteritems()}
+
+                    responses = []
+                    for key, group_id, _ in tasks:
+                        session.groups = [group_id]
+                        session.timeout = timeouts[group_id]
+                        responses.append(session.remove(key))
+
+                    failed_tasks = []
+                    is_last_attempt = attempt == ctx.attempts - 1
+                    for i, r in enumerate(responses):
+                        key, group_id, _ = tasks[i]
+                        status = r.get()[0].status
+                        log.info('Removed uncommitted key: %s, group: %s, status: %s, last attempt: %s',
+                                 key, group_id, status, is_last_attempt)
```

Comment:
`is_last_attempt` is less informative then `{attempt}/{attempts}`, can we use last variant?